Fix dependencies and bump bluepysnap

GianlucaFicarelli opened the pull request:
Fix dependencies in:
- py-bluepysnap
- py-bluepymm

Add ^neuron%intel to:
- py-minis-validation
- py-morph-tool

Bump:
- py-bluepysnap to 0.10.0

GianlucaFicarelli committed:
Fix dependencies and bump bluepysnap

Fix dependencies in:
- py-bluepysnap
- py-bluepymm
- py-neurodamus

Bump:
- py-bluepysnap to 0.10.0

GianlucaFicarelli committed:
Add ^neuron%intel to py-neurodamus

GianlucaFicarelli commented on deploy/environments/applications.yaml:
```diff
@@ -58,7 +58,7 @@ spack:
     - py-morphology-repair-workflow
     - py-morph-tool
-    - py-neurodamus
+    - py-neurodamus ^neuron%intel
```
I think this is needed as in https://github.com/BlueBrain/spack/pull/1066, but please confirm.

GianlucaFicarelli commented on deploy/environments/applications.yaml:
```diff
@@ -58,7 +58,7 @@ spack:
     - py-morphology-repair-workflow
     - py-morph-tool
-    - py-neurodamus
+    - py-neurodamus%gcc ^neuron%intel
```
I think this is needed as in https://github.com/BlueBrain/spack/pull/1066, but please confirm if it's correct.

pramodk replied:
yes - if a) package is compiled with non-intel compiler and 2) you are adding neuron as dependency

matz-e replied:
@pramodk this is dang-r-ous. see that other build failure. if we have overlapping dependencies, this may cause other parts of the DAG to be compiled with Intel and fail :(

pramodk replied:
As discussed on Slack, we should avoid neuron dependency on py-neurodamus.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-neurodamus/package.py:
```diff
@@ -33,6 +33,7 @@ class PyNeurodamus(PythonPackage):
     # We depend on Neurodamus but let the user decide which one
     depends_on('python@3.4:',      type=('build', 'run'))
     depends_on('py-setuptools',    type=('build', 'run'))
+    depends_on('neuron',           type='run')
```
Thinking about it, maybe `neuron` was not loaded intentionally, and the previous comment should be changed to
> _We depend on NEURON but let the user decide which one_

? Who can confirm?

pramodk replied:
@ferdonline is right person for this to answer.

(on BB5 deployment we typically have a single neuron installation and hence adding neuron here is not harmless)

matz-e replied:
I'll leave this for @ferdonline , from my side the rest of this PR is good.

pramodk replied:
I answered question about this on HPC slack : due to issues with Intel and GCC conflicts, we don't suggest to add this dependency.

mgeplf replied:
> due to issues with Intel and GCC conflicts, we don't suggest to add this dependency

Can this impact us when people load both packages?  ie: will we be getting an incompatible mixture of intel & gcc compiled code?

asanin-epfl commented on var/spack/repos/builtin/packages/py-bluepymm/package.py:
```diff
@@ -26,6 +26,8 @@ class PyBluepymm(PythonPackage):
     depends_on('py-lxml', type='run')
     depends_on('py-sh', type='run')
     depends_on('neuron', type='run')
+    depends_on('py-h5py', type='run')
```
we need to restrict version here `depends_on('py-h5py@:2.99', type='run')` because bluepymm still uses old API to access strings https://github.com/BlueBrain/BluePyMM/blob/bcd866f9af7f33c88fc82838ada64fd08add93cb/bluepymm/prepare_combos/parse_files.py#L213.

GianlucaFicarelli replied:
Thanks, I restricted the dependency to h5py < 3.
@anilbey @wvangeit could you confirm if you are fine with these new dependencies in `py-bluepymm` (`py-h5py`, `py-pyyaml`) that I took from `setup.py` [here](https://github.com/BlueBrain/BluePyMM/blob/bcd866f9af7f33c88fc82838ada64fd08add93cb/setup.py#L30-L31)?
Should you need to restrict to h5py < 3 also in `setup.py`?

anilbey replied:
BluePyMM works with both h5py<3 and h5py>=3. 
With this commit below if string is represented as bstr it's decoding it otherwise passing the original str. https://github.com/BlueBrain/BluePyMM/commit/b16e1652c8504bd5bf00cf6db14e9c025aebbfb8

asanin-epfl replied:
Sorry @GianlucaFicarelli . I missed this post-processing. It should work with both versions.

GianlucaFicarelli committed:
py-bluepymm should depend on py-h5py<3

anilbey commented on var/spack/repos/builtin/packages/py-bluepymm/package.py:
```diff
@@ -26,6 +26,8 @@ class PyBluepymm(PythonPackage):
+    depends_on('py-pyyaml', type='run')
 
     def setup_run_environment(self, env):
         env.unset('PMI_RANK')
```
Thanks for the update. The h5py version can also be >=3.0 for bluepymm.

GianlucaFicarelli replied:
Thanks, I reverted the last commit.

GianlucaFicarelli committed:
Revert "py-bluepymm should depend on py-h5py<3"

This reverts commit efd87e5dd49640a91fdeeb44a0128984541a2795.

GianlucaFicarelli committed:
Add ^neuron%intel to py-morph-tool and py-minis-validation

GianlucaFicarelli committed:
Remove neuron dependency from py-neurodamus

GianlucaFicarelli commented:
- I rolled back the changes to `py-neurodamus` (adding just a comment about `neuron`)
- I added `^neuron%intel` to `py-minis-validation` and `py-morph-tool` (that were already depending on neuron)

I also updated the PR description and now the CI is passing, @matz-e @pramodk could you have a look after the latest changes? Thanks.

GianlucaFicarelli commented:
@pramodk I think that the pending changes are fixed here (unless some packages are going to fail because trying to use the intel compiler for freetype, but the CI looks good).
Is there anything else that should be done in this PR?

pramodk reviewed (approved):
LGTM.

(As mentioned in the BSD JIRA ticket, one simplification could be we remove neuron dependency from packages and ask user to load neuron module. Adding one package to `module load` is not that much overhead)

GianlucaFicarelli commented:
Thanks, I'm going to merge this one.
